src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageClientMiddleLayer.cs

```diff
@@ -0,0 +1,215 @@
+                        vsTextEdit.Insert(span.Start, textEdit.NewText);
+                    }
+                }
+                else if (string.IsNullOrEmpty(textEdit.NewText))
+                {
+                    var startPosition = GetSnapshotPositionFromProtocolPosition(snapshot, textEdit.Range.Start);
+                    var endPosition = GetSnapshotPositionFromProtocolPosition(snapshot, textEdit.Range.End);
+                    var difference = endPosition - startPosition;
+                    if (startPosition > -1 && endPosition > -1 && difference > 0)
+                    {
+                        var span = GetTranslatedSpan(startPosition, difference, snapshot, vsTextEdit.Snapshot);
+                        vsTextEdit.Delete(span);
+                    }
+                }
+                else
+                {
+                    var startPosition = GetSnapshotPositionFromProtocolPosition(snapshot, textEdit.Range.Start);
+                    var endPosition = GetSnapshotPositionFromProtocolPosition(snapshot, textEdit.Range.End);
+                    var difference = endPosition - startPosition;
+
+                    if (startPosition > -1 && endPosition > -1 && difference > 0)
+                    {
+                        var span = GetTranslatedSpan(startPosition, difference, snapshot, vsTextEdit.Snapshot);
+                        vsTextEdit.Replace(span, textEdit.NewText);
+                    }
+                }
+            }
+
+            vsTextEdit.Apply();
+        }
+
+        private static Span GetTranslatedSpan(int startPosition, int length, ITextSnapshot oldSnapshot, ITextSnapshot newSnapshot)
+        {
+            var span = new Span(startPosition, length);
+
+            if (oldSnapshot.Version != newSnapshot.Version)
+            {
+                var snapshotSpan = new SnapshotSpan(oldSnapshot, span);
+                var translatedSnapshotSpan = snapshotSpan.TranslateTo(newSnapshot, SpanTrackingMode.EdgeExclusive);
+                span = translatedSnapshotSpan.Span;
+            }
+
+            return span;
+        }
+
+        internal static SnapshotPoint GetSnapshotPositionFromProtocolPosition(ITextSnapshot textSnapshot, Position position)
+        {
+            var line = textSnapshot.GetLineFromLineNumber(position.Line);
+            var snapshotPosition = line.Start + position.Character;
+
+            return new SnapshotPoint(textSnapshot, snapshotPosition);
+        }
+
+        internal static void MoveCaretToPosition(IVsTextView textView, Position position, bool sendFocus = true)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            textView.SetCaretPos(position.Line, position.Character);
+            textView.EnsureSpanVisible(new TextSpan() { iStartIndex = position.Character, iStartLine = position.Line, iEndIndex = position.Character, iEndLine = position.Line });
+            textView.CenterLines(position.Line, 1);
+            if (sendFocus)
```

Comment:
Interesting, where'd this piece come from?

Reply:
https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=%2Fsrc%2Fproduct%2FRemoteLanguage%2FImpl%2FUtilities.cs